x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/perform_bulk_action_route.ts

```diff
@@ -86,93 +81,54 @@ const normalizeErrorResponse = (errors: RuleActionError[]): NormalizedRuleError[
-  const errorsCount = errors.length;
-  return {
-    message: errorsCount === rulesCount ? 'Bulk edit failed' : 'Bulk edit partially failed',
-    status_code: 500,
-    attributes: {
-      errors: normalizeErrorResponse(errors).map(({ message, ...error }) => ({
-        ...error,
-        message:
-          message.length > MAX_ERROR_MESSAGE_LENGTH
-            ? `${message.slice(0, MAX_ERROR_MESSAGE_LENGTH - 3)}...`
-            : message,
-      })),
-      rules: {
-        total: rulesCount,
-        failed: errorsCount,
-        succeeded: rulesCount - errorsCount,
-      },
-    },
+const buildBulkResponse = (
+  response: KibanaResponseFactory,
+  fetchRulesOutcome: PromisePoolOutcome<string, RuleAlertType>,
+  bulkActionOutcome: PromisePoolOutcome<RuleAlertType, RuleAlertType | null>
+) => {
+  const errors = [...fetchRulesOutcome.errors, ...bulkActionOutcome.errors];
+  const summary = {
+    failed: errors.length,
+    succeeded: bulkActionOutcome.results.length,
+    total: bulkActionOutcome.results.length + errors.length,
   };
-};
 
-const executeActionAndHandleErrors = async (
-  rule: Rule,
-  action: RuleActionFn
-): Promise<RuleActionResult> => {
-  try {
-    await action(rule);
-  } catch (err) {
-    const { message, statusCode } = transformError(err);
-    return {
-      error: { message, statusCode },
-      rule: { id: rule.id, name: rule.name },
-    };
-  }
-};
-
-const executeBulkAction = async (rules: Rule[], action: RuleActionFn, abortSignal: AbortSignal) =>
-  initPromisePool({
-    concurrency: MAX_RULES_TO_UPDATE_IN_PARALLEL,
-    items: rules,
-    executor: async (rule) => executeActionAndHandleErrors(rule, action),
-    abortSignal,
-  });
-
-const getRulesByIds = async ({
-  ids,
-  rulesClient,
-  isRuleRegistryEnabled,
-  abortSignal,
-}: {
-  ids: string[];
-  rulesClient: RulesClient;
-  isRuleRegistryEnabled: boolean;
-  abortSignal: AbortSignal;
-}) => {
-  const readRulesExecutor = async (id: string) => {
-    try {
-      const rule = await readRules({ id, rulesClient, isRuleRegistryEnabled, ruleId: undefined });
-      if (rule == null) {
-        throw Error('Can`t fetch a rule');
-      }
-      return { rule };
-    } catch (err) {
-      const { message, statusCode } = transformError(err);
-      return {
-        error: { message, statusCode },
-        rule: { id },
-      };
-    }
+  const results = {
+    updated: bulkActionOutcome.results
+      .filter(({ item, result }) => item.id === result?.id)
+      .map(({ result }) => result && internalRuleToAPIResponse(result)),
+    created: bulkActionOutcome.results
+      .filter(({ item, result }) => result != null && result.id !== item.id)
+      .map(({ result }) => result && internalRuleToAPIResponse(result)),
+    deleted: bulkActionOutcome.results
+      .filter(({ result }) => result == null)
+      .map(({ item }) => internalRuleToAPIResponse(item)),
   };
```

Comment:
this logic looks a bit unobvious to me, plus, can be there situations, when rule gets in any both properties? For example in  deleted & updated?

I would rather have some excluding condition:

```
if (isUpdated) {
// add to updated
} else if (ifCreated) {
// add to created
} else if(isDeleted) {
// add toDeleted
}

```

or even more robust approach(I would prefer this one personally):
- take deleted array only from BulkAction.delete action result,
- created only from BulkAction.duplicate
- updated from the rest

this way, we won't need any additional logic for parsing results + it will ensure rule will be added to only one list

Reply:
> this logic looks a bit unobvious to me, plus, can be there situations, when rule gets in any both properties? For example in deleted & updated?

Not sure I understand you, how a rule could be updated and deleted at the same time?

Reply:
```
    updated: bulkActionOutcome.results
      .filter(({ item, result }) => item.id === result?.id)
      .map(({ result }) => result && internalRuleToAPIResponse(result)),
    created: bulkActionOutcome.results
      .filter(({ item, result }) => result != null && result.id !== item.id)
      .map(({ result }) => result && internalRuleToAPIResponse(result)),
    deleted: bulkActionOutcome.results
      .filter(({ result }) => result == null)
      .map(({ item }) => internalRuleToAPIResponse(item)),
```

if item.id === undefined and result === null, you will end up with rule in both updated and deleted arrays.
That's why I would prefer more logic that would exclude such possibilities

Reply:
> Not sure I understand you, how a rule could be updated and deleted at the same time?

Rule can't be, indeed. But because, we multiple times map through `bulkActionOutcome` could be there situations when not only single, but few properties get populated out of three(deleted, created, updated). 

In my opinion, it would be more readable and robust per traverse of a single time add it on of the arrays. 
Or populate arrays from particular actions only:

> take deleted array only from BulkAction.delete action result,
created only from BulkAction.duplicate
updated from the rest

Reply:
The current logic is pretty straightforward:

- We had a rule, now it's null - the rule was deleted
- We had a rule, now there's a rule with a different id - a new rule was created
- We had a rule, now there's a rule with the same id - the existing rule was modified

It doesn't depend on our business logic, so there's no need to tightly couple this with bulk actions whatsoever. Instead, I prefer these pieces to be separated.